lib/IOExt/IOExt.cpp

```diff
@@ -3,253 +3,219 @@
+    arrow_decrease(true);
+  } else {
+    arrow_decrease(false);
+  }
+}
 
-    // int v[8] = {0, 0, 0, 0, 0, 0, 0, 0};
-    // for (int idx = 0; idx < 8; idx++)
-    // {
-    //     v[idx] = pcf8574.digitalRead(idx);
-    // }
-    // for (int idx = 0; idx < 8; idx++)
-    // {
-    //     printf("%d:%d - ", idx, v[idx]);
-    // }
-    // Serial.println();
+void _handleIoInterrupt() {
+  // CRITICAL SECTION I2C: start
+  xSemaphoreTake(i2c_mutex, portMAX_DELAY);
+
+  PCF8574::DigitalInput dra = pcf8574.digitalReadAll();
+
+  xSemaphoreGive(i2c_mutex);
+  // CRITICAL SECTION I2C: end
+
+  if ((dra.p1 & dra.p2 & dra.p3 & dra.p4 & dra.p5 & dra.p6 & dra.p7) == 1) {
+    return;
+  }
+  printf("PCF: %d %d %d %d - %d %d %d %d\n", dra.p0, dra.p1, dra.p2, dra.p3,
+         dra.p4, dra.p5, dra.p6, dra.p7);
+
+  taskSleep = 10;
+
+  // turn indicator and hazard lights
+  if (!dra.p7 || !dra.p1) {
+    taskSleep = 200;
+    update_indicator(!dra.p7, !dra.p1);
+  }
+  // position lights on/off
+  if (!dra.p6) {
+    taskSleep = 100;
+    light1OnOff();
+  }
+  // head lights on/off
+  if (!dra.p5) {
+    taskSleep = 100;
+    light2OnOff();
+  }
+
+  // Simulation
+  if (!dra.p4) {
+    speed += 10;
+    write_speed(speed);
+    _speedCheck(speed);
+  }
+  if (!dra.p3) {
+    speed -= 10;
+    if (speed < 0) {
+      speed = 0;
+    }
+    write_speed(speed);
+    _speedCheck(speed);
+  }
+
+  // int v[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+  // for (int idx = 0; idx < 8; idx++)
+  // {
+  //     v[idx] = pcf8574.digitalRead(idx);
+  // }
+  // for (int idx = 0; idx < 8; idx++)
+  // {
+  //     printf("%d:%d - ", idx, v[idx]);
+  // }
+  // Serial.println();
 }
 
-void set_ioext(int port, bool value)
-{
-    // check port
-    if (port < 0 || port >= PCF8574_NUM_PORTS)
-    {
-        return;
-    }
+void set_ioext(int port, bool value) {
+  // check port
+  if (port < 0 || port >= PCF8574_NUM_PORTS) {
+    return;
+  }
 
-    // CRITICAL SECTION I2C: start
-    xSemaphoreTake(i2c_mutex, portMAX_DELAY);
+  // CRITICAL SECTION I2C: start
+  xSemaphoreTake(i2c_mutex, portMAX_DELAY);
 
-    pcf8574.digitalWrite(port, value);
+  pcf8574.digitalWrite(port, value);
 
-    xSemaphoreGive(i2c_mutex);
-    // CRITICAL SECTION I2C: end
+  xSemaphoreGive(i2c_mutex);
+  // CRITICAL SECTION I2C: end
 }
 
-int get_ioext(int port)
-{
-    // check port
-    if (port < 0 || port >= PCF8574_NUM_PORTS)
-    {
-        return -1;
-    }
+int get_ioext(int port) {
+  // check port
+  if (port < 0 || port >= PCF8574_NUM_PORTS) {
+    return -1;
+  }
 
-    // CRITICAL SECTION I2C: start
-    xSemaphoreTake(i2c_mutex, portMAX_DELAY);
+  // CRITICAL SECTION I2C: start
+  xSemaphoreTake(i2c_mutex, portMAX_DELAY);
 
-    int value = pcf8574.digitalRead(P0);
+  int value = pcf8574.digitalRead(P0);
 
-    xSemaphoreGive(i2c_mutex);
-    // CRITICAL SECTION I2C: end
+  xSemaphoreGive(i2c_mutex);
+  // CRITICAL SECTION I2C: end
 
-    return value;
+  return value;
 }
 
 int counter = 0;
```

Comment:
Can  you move the counter inside the fuction body?

Reply:
completely removed